test/unit/test_matrix.cpp

```diff
@@ -75,23 +126,61 @@ TYPED_TEST(MatrixTest, Constructor) {
   }
 }
 
-/// @brief Returns the memory index of the @p index element of the matrix.
-/// @pre index should be a valid and contained in @p layout.size().
-std::size_t memoryIndex(const LayoutInfo& layout, const GlobalElementIndex& index) {
+TYPED_TEST(MatrixTest, ConstructorFromDistribution) {
+  using Type = TypeParam;
+  auto el = [](const GlobalElementIndex& index) {
+    SizeType i = index.row();
+    SizeType j = index.col();
+    return TypeUtilities<Type>::element(i + 0.001 * j, j - 0.01 * i);
+  };
+
+  for (const auto& comm_grid : this->commGrids()) {
+    for (const auto& test : sizes_tests) {
+      GlobalElementSize size(test.size);
+      comm::Index2D src_rank_index(std::max(0, comm_grid.size().rows() - 1),
+                                   std::min(1, comm_grid.size().cols() - 1));
+      Distribution distribution(size, test.block_size, comm_grid.size(), comm_grid.rank(),
+                                src_rank_index);
+
+      // Copy distribution for testing purpose.
+      Distribution distribution_copy(distribution);
+
+      Matrix<Type, Device::CPU> mat(std::move(distribution));
+
+      EXPECT_EQ(distribution_copy, mat.distribution());
+
+      set(mat, el);
+
+      CHECK_MATRIX_EQ(el, mat);
+    }
+  }
+}
+
+/// Returns the memory index of the @p index element of the matrix.
+///
+/// @pre index should be valid, contained in @p distribution.size() and stored in the current rank.
+std::size_t memoryIndex(const Distribution& distribution, const LayoutInfo& layout,
+                        const GlobalElementIndex& index) {
   using util::size_t::sum;
   using util::size_t::mul;
-  const auto& block_size = layout.blockSize();
-  SizeType tile_i = index.row() / block_size.rows();
-  SizeType tile_j = index.col() / block_size.cols();
-  std::size_t tile_offset = layout.tileOffset({tile_i, tile_j});
-  SizeType i = index.row() % block_size.rows();
-  SizeType j = index.col() % block_size.cols();
-  std::size_t element_offset = sum(i, mul(layout.ldTile(), j));
+
+  auto global_tile_index = distribution.globalTileIndex(index);
+  auto tile_element_index = distribution.tileElementIndex(index);
+  auto local_tile_index = distribution.localTileIndex(global_tile_index);
+  std::size_t tile_offset = layout.tileOffset(local_tile_index);
+  std::size_t element_offset =
+      sum(tile_element_index.row(), mul(layout.ldTile(), tile_element_index.col()));
   return tile_offset + element_offset;
 }
+/// Returns true if the memory index is stored in distribution.rankIndex().
+bool ownIndex(const Distribution& distribution, const GlobalElementIndex& index) {
+  auto global_tile_index = distribution.globalTileIndex(index);
+  return distribution.rankIndex() == distribution.rankGlobalTile(global_tile_index);
+}
```

Comment:
Isn't this something that may be useful in general and not just for tests?

Reply:
No, as the note on line 206 warns, this is the most straightforward but very inefficient method to loop over elements.